File: ApexFit/Engines/StatisticalEngine.swift
import Foundation

struct CorrelationResult {
    let behaviorName: String
    let metricName: String
    let effectSize: Double      // Cohen's d
    let pValue: Double
    let isSignificant: Bool     // p < 0.05
    let direction: CorrelationDirection
    let sampleSizeWith: Int
    let sampleSizeWithout: Int
    let meanWith: Double
    let meanWithout: Double
}

enum CorrelationDirection: String {
    case positive   // Behavior improves the metric
    case negative   // Behavior worsens the metric
    case neutral    // No significant effect
}

struct StatisticalEngine {
    /// Perform independent samples t-test comparing metric values
    /// when a behavior is present vs absent.
    static func tTest(
        withBehavior: [Double],
        withoutBehavior: [Double]
    ) -> (tStatistic: Double, pValue: Double)? {
        guard withBehavior.count >= 3, withoutBehavior.count >= 3 else { return nil }

        let n1 = Double(withBehavior.count)
        let n2 = Double(withoutBehavior.count)
        let mean1 = withBehavior.mean
        let mean2 = withoutBehavior.mean
        let var1 = withBehavior.standardDeviation * withBehavior.standardDeviation
        let var2 = withoutBehavior.standardDeviation * withoutBehavior.standardDeviation

        let pooledSE = sqrt(var1 / n1 + var2 / n2)
        guard pooledSE > 0 else { return nil }

        let t = (mean1 - mean2) / pooledSE

        // Welch-Satterthwaite degrees of freedom
        let num = pow(var1 / n1 + var2 / n2, 2)
        let den = pow(var1 / n1, 2) / (n1 - 1) + pow(var2 / n2, 2) / (n2 - 1)
        let df = num / den

        // Approximate p-value using normal distribution for large samples
        let pValue = approximatePValue(t: abs(t), df: df)

        return (t, pValue)
    }

    /// Compute Cohen's d effect size.
    static func cohensD(
        withBehavior: [Double],
        withoutBehavior: [Double]
    ) -> Double? {
        guard withBehavior.count >= 3, withoutBehavior.count >= 3 else { return nil }

        let mean1 = withBehavior.mean
        let mean2 = withoutBehavior.mean
        let sd1 = withBehavior.standardDeviation
        let sd2 = withoutBehavior.standardDeviation
        let n1 = Double(withBehavior.count)
        let n2 = Double(withoutBehavior.count)

        // Pooled standard deviation
        let pooledSD = sqrt(((n1 - 1) * sd1 * sd1 + (n2 - 1) * sd2 * sd2) / (n1 + n2 - 2))
        guard pooledSD > 0 else { return nil }

        return (mean1 - mean2) / pooledSD
    }

    /// Full correlation analysis for a behavior against a metric.
    static func analyzeCorrelation(
        behaviorName: String,
        metricName: String,
        withBehavior: [Double],
        withoutBehavior: [Double],
        higherIsBetter: Bool = true
    ) -> CorrelationResult? {
        guard let testResult = tTest(withBehavior: withBehavior, withoutBehavior: withoutBehavior),
              let effectSize = cohensD(withBehavior: withBehavior, withoutBehavior: withoutBehavior) else {
            return nil
        }

        let meanDiff = withBehavior.mean - withoutBehavior.mean
        let direction: CorrelationDirection
        if !testResult.pValue.isLess(than: 0.05) {
            direction = .neutral
        } else if (higherIsBetter && meanDiff > 0) || (!higherIsBetter && meanDiff < 0) {
            direction = .positive
        } else {
            direction = .negative
        }

        return CorrelationResult(
            behaviorName: behaviorName,
            metricName: metricName,
            effectSize: effectSize,
            pValue: testResult.pValue,
            isSignificant: testResult.pValue < 0.05,
            direction: direction,
            sampleSizeWith: withBehavior.count,
            sampleSizeWithout: withoutBehavior.count,
            meanWith: withBehavior.mean,
            meanWithout: withoutBehavior.mean
        )
    }

    /// Effect size interpretation.
    static func interpretEffectSize(_ d: Double) -> String {
        let absD = abs(d)
        switch absD {
        case ..<0.2: return "Negligible"
        case 0.2..<0.5: return "Small"
        case 0.5..<0.8: return "Medium"
        default: return "Large"
        }
    }

    // MARK: - Private

    /// Approximate two-tailed p-value using the normal distribution.
    private static func approximatePValue(t: Double, df: Double) -> Double {
        // For df > 30, t-distribution approximates normal
        // Using a simple approximation
        let x = abs(t)
        let p = erfc(x / sqrt(2.0))
        return p
    }
}

File: android/core/engine/src/main/kotlin/com/apexfit/core/engine/StatisticalEngine.kt
package com.apexfit.core.engine

import kotlin.math.abs
import kotlin.math.pow
import kotlin.math.sqrt

enum class CorrelationDirection {
    POSITIVE,  // Behavior improves the metric
    NEGATIVE,  // Behavior worsens the metric
    NEUTRAL,   // No significant effect
}

data class CorrelationResult(
    val behaviorName: String,
    val metricName: String,
    val effectSize: Double,
    val pValue: Double,
    val isSignificant: Boolean,
    val direction: CorrelationDirection,
    val sampleSizeWith: Int,
    val sampleSizeWithout: Int,
    val meanWith: Double,
    val meanWithout: Double,
)

object StatisticalEngine {

    /**
     * Perform independent samples Welch's t-test comparing metric values
     * when a behavior is present vs absent.
     */
    fun tTest(
        withBehavior: List<Double>,
        withoutBehavior: List<Double>,
    ): Pair<Double, Double>? {
        if (withBehavior.size < 3 || withoutBehavior.size < 3) return null

        val n1 = withBehavior.size.toDouble()
        val n2 = withoutBehavior.size.toDouble()
        val mean1 = withBehavior.mean
        val mean2 = withoutBehavior.mean
        val var1 = withBehavior.standardDeviation.pow(2)
        val var2 = withoutBehavior.standardDeviation.pow(2)

        val pooledSE = sqrt(var1 / n1 + var2 / n2)
        if (pooledSE <= 0) return null

        val t = (mean1 - mean2) / pooledSE

        // Approximate p-value using normal distribution for large samples
        val pValue = approximatePValue(abs(t))

        return t to pValue
    }

    /**
     * Compute Cohen's d effect size.
     */
    fun cohensD(
        withBehavior: List<Double>,
        withoutBehavior: List<Double>,
    ): Double? {
        if (withBehavior.size < 3 || withoutBehavior.size < 3) return null

        val mean1 = withBehavior.mean
        val mean2 = withoutBehavior.mean
        val sd1 = withBehavior.standardDeviation
        val sd2 = withoutBehavior.standardDeviation
        val n1 = withBehavior.size.toDouble()
        val n2 = withoutBehavior.size.toDouble()

        val pooledSD = sqrt(((n1 - 1) * sd1 * sd1 + (n2 - 1) * sd2 * sd2) / (n1 + n2 - 2))
        if (pooledSD <= 0) return null

        return (mean1 - mean2) / pooledSD
    }

    /**
     * Full correlation analysis for a behavior against a metric.
     */
    fun analyzeCorrelation(
        behaviorName: String,
        metricName: String,
        withBehavior: List<Double>,
        withoutBehavior: List<Double>,
        higherIsBetter: Boolean = true,
    ): CorrelationResult? {
        val testResult = tTest(withBehavior, withoutBehavior) ?: return null
        val effectSize = cohensD(withBehavior, withoutBehavior) ?: return null

        val meanDiff = withBehavior.mean - withoutBehavior.mean
        val direction = when {
            testResult.second >= 0.05 -> CorrelationDirection.NEUTRAL
            (higherIsBetter && meanDiff > 0) || (!higherIsBetter && meanDiff < 0) -> CorrelationDirection.POSITIVE
            else -> CorrelationDirection.NEGATIVE
        }

        return CorrelationResult(
            behaviorName = behaviorName,
            metricName = metricName,
            effectSize = effectSize,
            pValue = testResult.second,
            isSignificant = testResult.second < 0.05,
            direction = direction,
            sampleSizeWith = withBehavior.size,
            sampleSizeWithout = withoutBehavior.size,
            meanWith = withBehavior.mean,
            meanWithout = withoutBehavior.mean,
        )
    }

    fun interpretEffectSize(d: Double): String {
        val absD = abs(d)
        return when {
            absD < 0.2 -> "Negligible"
            absD < 0.5 -> "Small"
            absD < 0.8 -> "Medium"
            else -> "Large"
        }
    }

    /**
     * Approximate two-tailed p-value using the complementary error function.
     */
    private fun approximatePValue(t: Double): Double {
        val x = abs(t)
        return erfc(x / sqrt(2.0))
    }

    /**
     * Complementary error function approximation.
     * Uses Abramowitz and Stegun approximation (formula 7.1.26).
     */
    private fun erfc(x: Double): Double {
        val t = 1.0 / (1.0 + 0.3275911 * abs(x))
        val poly = t * (0.254829592 + t * (-0.284496736 + t * (1.421413741 + t * (-1.453152027 + t * 1.061405429))))
        val result = poly * kotlin.math.exp(-x * x)
        return if (x >= 0) result else 2.0 - result
    }
}
